### PR TITLE
feat(stream): support replace source reader at runtime

### PR DESCRIPTION
**Pull request by tabVersion**
Signed-off-by: tabVersion <tabvision@bupt.icu>

## What's changed and what's your intention?

the process for handling change_split mutation is like:

1. check if the barrier carries change_split mutation
2. fetch assigned splits by actor_id
3. check if assigned splits are all in the state cache. If yes, ignore the barrier. If no, rebuild source with new splits. Note that the states will use offsets from the cache if possible. 

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)
- Add the 'user-facing changes' label if your PR contains changes that are visible to users (optional) 

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)

part of #1285

### CODE REVIEW

**Comment by tabVersion** [comment]
> here we introduce a hypothesis that every split is polled at least once in one epoch.

please consider whether it is reasonable

File: src/compute/tests/table_v2_materialize.rs [hunk 1]
```diff
@@ -134,6 +134,7 @@ async fn test_table_v2_materialize() -> Result<()> {
     let (barrier_tx, barrier_rx) = unbounded_channel();
     let keyspace = Keyspace::executor_root(MemoryStateStore::new(), 0x2333);
     let stream_source = SourceExecutor::new(
+        0x3f3f3f,
```
Comment (neverchanje) [inline src/compute/tests/table_v2_materialize.rs:137]: pls use a constant var to represent this magic number.

**Review by BugenZhao — COMMENTED** [review]
The source part generally LGTM

File: src/stream/src/executor/source.rs [hunk 2]
```diff
@@ -150,12 +153,16 @@ impl SourceReader {
...
         notifier: Arc<Notify>,
         expected_barrier_latency_ms: u64,
+        mut inject_source: UnboundedReceiver<Box<dyn StreamSourceReader>>,
     ) {
         'outer: loop {
             let now = Instant::now();
 
             // We allow data to flow for `expected_barrier_latency_ms` milliseconds.
             while now.elapsed().as_millis() < expected_barrier_latency_ms as u128 {
+                if let Some(reader) = inject_source.recv().await {
```
Comment (shanicky) [inline src/stream/src/executor/source.rs:163]: Is it possible to be blocked here?

**Review by shanicky — APPROVED** [review]
rest LGTM

File: src/stream/src/executor/source.rs [hunk 3]
```diff
@@ -244,11 +281,28 @@ impl<S: StateStore> SourceExecutor<S> {
...
+                .stream_reader(self.column_ids.clone())
+                .await
+                .map(SourceStreamReaderImpl::TableV2),
+            SourceImpl::Connector(c) => c
+                .stream_reader(state, self.column_ids.clone())
+                .await
+                .map(SourceStreamReaderImpl::Connector),
+        }?;
+
+        Ok(Box::new(reader))
```
[the remark this answers is not in the document]
Comment (tabVersion) [reply src/stream/src/executor/source.rs:303]: fixed

**Pull request merged**